controllers/humiocluster_controller.go

```diff
@@ -1168,6 +1173,74 @@ func (r *HumioClusterReconciler) ensurePvcLabels(ctx context.Context, hc *humiov
 	return nil
 }
 
+func (r *HumioClusterReconciler) ensureLicense(ctx context.Context, hc *humiov1alpha1.HumioCluster) (reconcile.Result, error) {
+	r.Log.Info("ensuring license")
+
+	existingLicense, err := r.HumioClient.GetLicense()
+	if err != nil {
+		r.Log.Info(fmt.Sprintf("failed to get license: %v", err))
+	}
+
+	defer func(ctx context.Context, hc *humiov1alpha1.HumioCluster) {
+		licenseStatus := humiov1alpha1.HumioLicenseStatus{
+			Type:       existingLicense.LicenseType(),
+			Expiration: existingLicense.ExpiresAt(),
+		}
+		r.setLicense(ctx, licenseStatus, hc)
+	}(ctx, hc)
+
+	licenseSecretKeySelector := licenseSecretKeyRefOrDefault(hc)
+	if licenseSecretKeySelector == nil {
+		return reconcile.Result{}, nil
+	}
+
+	licenseSecret, err := kubernetes.GetSecret(ctx, r, licenseSecretKeySelector.Name, hc.Namespace)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Error(err, fmt.Sprintf("license was requested but no secret exists by name %s in namespace %s", licenseSecretKeySelector.Name, hc.Namespace))
+			return reconcile.Result{}, err
+		}
+	}
+	if _, ok := licenseSecret.Data[licenseSecretKeySelector.Key]; !ok {
+		r.Log.Error(err, fmt.Sprintf("license secret was found but it does not contain the key %s", licenseSecretKeySelector.Key))
+		return reconcile.Result{}, err
+	}
+
+	licenseBytes, err := base64.StdEncoding.DecodeString(string(licenseSecret.Data[licenseSecretKeySelector.Key]))
+	if err != nil {
+		r.Log.Error(err, fmt.Sprintf("license was supplied but could not be decoded %s", err))
+		return reconcile.Result{}, err
+	}
+	licenseStr := string(licenseBytes)
+
+	desiredLicense, err := humio.ParseLicense(licenseStr)
+	if err != nil {
+		r.Log.Error(err, fmt.Sprintf("license was supplied but could not be parsed %s", err))
+		return reconcile.Result{}, err
+	}
+
+	if existingLicense == nil || (existingLicense.LicenseType() != desiredLicense.LicenseType() ||
+		existingLicense.IssuedAt() != desiredLicense.IssuedAt() ||
+		existingLicense.ExpiresAt() != desiredLicense.ExpiresAt()) {
+		if err := r.HumioClient.InstallLicense(licenseStr); err != nil {
+			r.Log.Error(err, "could not install license")
+			return reconcile.Result{}, err
+		}
+
+		r.Log.Info(fmt.Sprintf("successfully installed license: type: %s, issued: %s, expires: %s",
+			desiredLicense.LicenseType(), desiredLicense.IssuedAt(), desiredLicense.ExpiresAt()))
+
+		// refresh the existing license for the status update
+		existingLicense, err = r.HumioClient.GetLicense()
+		if err != nil {
+			r.Log.Error(err, "failed to get updated license: %v", err)
+		}
```

Comment:
It seems slightly weird to me that we have these lines here, given that they are not important for ensuring the license is installed to the Humio cluster. I feel like they should be put closer to the code that actually updates the license status. Right now it is not clear when reading these few lines, what other parts of the code depends on it.

Reply:
I agree it would be nice to clean up the main Reconcile function. There are a number of status update helper functions that are not as easily correlated to an ensure method like license is. But perhaps we should do some refactoring in this area.